### PR TITLE
2 add a value feature to items

### LINKED ISSUE #2: Add a value feature to items, update reward and generator to accommodate for this
### Is your feature request related to a problem? Please describe
One industrial application of a single bin bin packing problem is to maximise the value packed rather than the volume. This value could be in euros for example.

### Describe the solution you'd like
Add a value feature to items and placed items.
Include item value in the generator.
Have a value based reward option.

### Describe alternatives you've considered
No alternatives considered.

### Additional context

---
### Misc
- [ ] Check for duplicate requests.

### PR DESCRIPTION
**Pull request by rsjeffers**
This PR resolves this issue: #2

Changes in this PR:

- Add a ValuedItem type to types which is the same as Item but with an added "value" feature
- New RandomValueProblemGenerator and ValueProblemCSVGenerator generators for such problems
- New rewards: ValueBasedDenseReward and ValueBasedSparseReward
- Modified specs in order to use either "Item" or "ValuedItem"
- alias for Union[Item, ValuedItem] called ItemType
- The ability to load an environment from config instead of registered environments
- example value-based training script with toy configs

### CODE REVIEW

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 1]
```diff
@@ -79,6 +89,55 @@ def __call__(self, key: chex.PRNGKey) -> State:
         )
 
 
+class DummyValueGenerator(Generator):
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/conftest.py:95]: Avoid code duplication by inheriting directly from `DummyGenerator` and only modifying the items attribute of State inside the `__call__` function.

File: jumanji/environments/packing/bin_pack/reward.py [hunk 2]
```diff
@@ -91,3 +92,62 @@ def sparse_reward(state: State) -> jnp.float_:
...
+        action: chex.Array,
+        next_state: State,
+        is_valid: bool,
+        is_done: bool,
+    ) -> float:
+        del next_state, is_done
+        state.items = cast(ValuedItem, state.items)
+        _, item_id = action
+        chosen_item_value = tree_slice(state.items, item_id).value
+        instance_total_value = jnp.sum(state.items.value * state.items_mask)
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/reward.py]: This can be stored as an attribute of the state, in case `items_mask` changes during the episode.

File: jumanji/environments/packing/bin_pack/reward.py [hunk 3]
```diff
@@ -91,3 +92,62 @@ def sparse_reward(state: State) -> jnp.float_:
...
+        next_state: State,
+        is_valid: bool,
+        is_done: bool,
+    ) -> float:
+        del state, action, is_valid
+
+        def sparse_reward(state: State) -> jnp.float_:
+            """Returns volume utilization between 0.0 and 1.0."""
+            state.items = cast(ValuedItem, state.items)
+            items_value = jnp.sum(state.items.value * state.items_placed)
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/reward.py:142]: Same comment as for `ValueBasedDenseReward`

File: jumanji/environments/packing/bin_pack/generator_test.py [hunk 4]
```diff
@@ -203,3 +208,87 @@ def test_random_generator__generate_solution(
...
+        # This will produce a solution to the environment (all possible items packed).
+        solution_state1 = generate_solution(jax.random.PRNGKey(1))
+        assert isinstance(solution_state1, State)
+        assert_trees_are_equal(solution_state1.ems, state1.ems)
+        # Should be different because there is only 1 available ems in state1, whereas there should
+        # be none available at the end with a solution.
+        assert_trees_are_different(solution_state1.ems_mask, state1.ems_mask)
+        # The items do not change whether they are packed or not, it is just the items-placed mask
+        # that is different.
+        assert_trees_are_close(solution_state1.items, state1.items)
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/generator_test.py:267]: Why are you using `assert_trees_are_close` instead of `assert_trees_are_equal`, is it because of the dimension normalization, I think it is worth mentioning if that's the case.

File: jumanji/environments/packing/bin_pack/generator_test.py [hunk 5]
```diff
@@ -203,3 +208,87 @@ def test_random_generator__generate_solution(
...
+        # This is because the solution is generated first, in a real solution the items mask would
+        # be all false at the end.
+        assert_trees_are_equal(solution_state1.items_mask, state1.items_mask)
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/generator_test.py:268]: The `items_mask` doesn't appear to be changing currently, so I think this comment is a bit confusing.

File: jumanji/environments/packing/bin_pack/generator_test.py [hunk 6]
```diff
@@ -203,3 +208,87 @@ def test_random_generator__generate_solution(
...
+            item_volume(solution_state1.items) * solution_state1.items_placed
+        ).sum()
+        assert jnp.isclose(placed_items_volume, solution_state1.container.volume())
+
+        # Checks that the total volume of all items in an instance is greater than the volume of the
+        # container
+        items_volume = (
+            item_volume(solution_state1.items) * solution_state1.items_mask
+        ).sum()
+        assert not jnp.isclose(items_volume, solution_state1.container.volume())
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/generator_test.py:288]: `jnp.isclose` might return True if the item_volume is lower than or equal to the container volume though.

File: jumanji/environments/packing/bin_pack/env.py [hunk 7]
```diff
@@ -211,19 +219,36 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
         else:
-            items_dict = {
-                f"{axis}": specs.BoundedArray(
-                    (max_num_items,), jnp.int32, 0, max_dim, axis
-                )
-                for axis in ["x_len", "y_len", "z_len"]
-            }
-        items = specs.Spec(Item, "ItemsSpec", **items_dict)
+            if valued_items_used:
+                items_dict = {
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/env.py]: Why aren't you keeping the `x_len`, `y_len`, `z_len` integers here ?

File: jumanji/environments/packing/bin_pack/env.py [hunk 8]
```diff
@@ -211,19 +219,36 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
+            else ["x_len", "y_len", "z_len"]
+        )
         if self.normalize_dimensions:
             items_dict = {
                 f"{axis}": specs.BoundedArray((max_num_items,), float, 0.0, 1.0, axis)
-                for axis in ["x_len", "y_len", "z_len"]
+                for axis in items_axes
             }
         else:
-            items_dict = {
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/env.py:220]: I suggest you keep this and just update the items_dict with the specs of the value attribute.

File: jumanji/environments/packing/bin_pack/env.py
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/env.py:1]: I think the checks that you're applying in each function in order to see whether you're dealing with an instance with values or not, is a good enough reason to create a new environment that inherits from `BinPack` where you redefine the necessary attributes and methods. This would improve code readability and allow you to reduce the number of checks that you have to do inside each method.

File: jumanji/environments/packing/bin_pack/reward_test.py
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/reward_test.py:1]: I think the new tests have a pretty good coverage although it would be interesting to add a test that checks for the value of the dense reward in the middle of an episode. Also, I think these tests might be falling if you change the structure of the reward to use the optimal solution instead of the sum of values of all the items. A quick fix would be to have a global variable or a class variable where you store this value.

File: jumanji/environments/packing/bin_pack/generator.py [hunk 9]
```diff
@@ -875,3 +895,330 @@ def body_fn(
...
+        ems_mask = jnp.zeros(self.max_num_ems, bool)
+
+        # Create less than half of max_num_items item spaces by splitting up a container. This
+        # will lead to nb_items_in_one_container items that fit perfectly into a single container.
+        nb_items_in_one_container = math.floor(0.5 * self.max_num_items)
+        # This will generate items_spaces and item_mask of size nb_items_in_one_container.
+        items_spaces, items_mask = self._split_container_into_items_spaces(
+            container, split_key, nb_items_in_one_container
+        )
+        # Randomly generate values that will then be doubled for the above items
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/generator.py]: I think the second part of this comment is outdated

File: jumanji/environments/packing/bin_pack/generator.py [hunk 10]
```diff
@@ -875,3 +895,330 @@ def body_fn(
...
+
+        # Create the solution state by creating trees of size self.max_num_items for items,
+        # items_placable_at_beginning_mask and items_placed_mask.
+        items = jax.tree_map(
+            lambda x, y, z: jnp.concatenate((x, y, z)),
+            optimal_items,
+            extra_items,
+            padding_items,
+        )
+        items_placable_at_beginning_mask = jnp.concatenate(
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/generator.py:1052]: why can't the padded item be placed at the beginning?

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 1 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/conftest.py:95]: I'd thought about that but the current implementation avoids mypy issues in overwriting items and we'd end up writing about the same amount of code anyway?

File: jumanji/environments/packing/bin_pack/env.py
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py:1]: Inheritance would be difficult from a typing point of view, maybe a wrapper ? Given that there aren't a large number of checks I'd prefer to leave it like this for this PR and then when we merge it with the other options think of a good way of combining all options?

File: jumanji/environments/packing/bin_pack/env.py (hunk 7 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py]: Good point, I should be able to change this but it will involve more code

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 4 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator_test.py:267]: Because value is a float

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 6 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator_test.py:288]: I'll add another check

File: jumanji/environments/packing/bin_pack/reward_test.py
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/reward_test.py:1]: I don't understand the second point?

File: jumanji/environments/packing/bin_pack/reward.py (hunk 2 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/reward.py]: Fixed in this commit a0c0a54ba34aeea1970da211fa9c414894940397

File: jumanji/environments/packing/bin_pack/reward.py (hunk 3 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/reward.py:142]: Fixed in this commit https://github.com/rsjeffers/jumanji_value_based/commit/a0c0a54ba34aeea1970da211fa9c414894940397

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 5 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator_test.py:268]: Removed in this commit dae30bfdf8a4e262740bc634f2237ed02fd88dc6

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 6 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator_test.py:288]: Fixed in this commit 29b0d5325a0ad430d8cf188be73415d9c2771f69

File: jumanji/environments/packing/bin_pack/env.py (hunk 7 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py]: Actually there's another error here, even if I normalise with the max value it won't be bounded at zero due to values being able to be negative - I'll refactor this tomorrow

File: jumanji/environments/packing/bin_pack/generator.py (hunk 9 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator.py]: Good spot! Updated here e3b9d44c46336ab2590874c1e969a51301bab5da

File: jumanji/environments/packing/bin_pack/generator.py (hunk 10 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator.py:1052]: Because they don't exist

File: jumanji/environments/packing/bin_pack/env.py
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:1]: Okay, I'm resolving this for now then.

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 1 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py:95]: Alright.

File: jumanji/environments/packing/bin_pack/env.py [hunk 11]
```diff
@@ -211,19 +219,36 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
         ems = specs.Spec(EMS, "EMSSpec", **ems_dict)
         ems_mask = specs.BoundedArray((obs_num_ems,), bool, False, True, "ems_mask")
+        items_axes = (
+            ["x_len", "y_len", "z_len", "value"]
+            if valued_items_used
+            else ["x_len", "y_len", "z_len"]
+        )
         if self.normalize_dimensions:
             items_dict = {
                 f"{axis}": specs.BoundedArray((max_num_items,), float, 0.0, 1.0, axis)
```
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/env.py]: You might get the same bug as the one you're speaking of [here](https://github.com/rsjeffers/jumanji_value_based/pull/4/files/5112be6fa8d3a16bd220236e49e6a93b6bd399be#r1267067438), so the value spec needs to be between -1 and 1 instead of 0 and 1.

File: jumanji/environments/packing/bin_pack/reward_test.py
Comment (chouakifares) [inline jumanji/environments/packing/bin_pack/reward_test.py:1]: Update the `instance_total_value` and `instance_max_value` variables to use the attributes of the State.

File: jumanji/environments/packing/bin_pack/env.py (hunk 8 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py:220]: Done as part of the refactoring.

File: jumanji/environments/packing/bin_pack/env.py (hunk 11 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py]: This is solved in the latest refactoring.

File: jumanji/environments/packing/bin_pack/reward_test.py
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/reward_test.py:1]: Done

Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/reward_test.py:1]: I'm closing this we'll talk about it more irl.

File: jumanji/environments/packing/bin_pack/env.py (hunk 7 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py]: Done

**Review by chouakifares — APPROVED** [review]
